### PR TITLE
[Feature] Zigzag Ring attention

### PR DESCRIPTION
**Pull request by Edenzzzz**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
1. Implements Ring Attention from https://arxiv.org/abs/2310.01889 and supports Llama. Adopts the zigzag batch splitting scheme for load-balancing (see illustration below).
image 

2. Supports both batched sequences and packed sequences.
3. Use two streams in forward pass to overlap the low-occupancy output correction with next step's flash attn kernel. Carefully optimizes buffers to minimize memory pressure. Provides a experimental Triton kernel for output correction
4. Save comm & compute by replacing the gather sequence op in cross entropy wtih reducing loss.
5. Enable TP + SP in plugin and update corresponding llama and command policies.

## ！Caveats/TODO
1. For easy compatibility with other SP and HF interfaces, in the packed sequence mode we do not flatten the batch dim but recreates a new tensor with shape (bs, max_seqlen // sp_size). To eliminate this overhead, we can adapt each model's forward to accept flattened inputs.
2. Pad each seq to be divisible by sp_size.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/layer/_operation.py [hunk 1]
```diff
@@ -804,7 +812,11 @@ def backward(ctx, *grad_output):
         process_group = ctx.process_group
         scatter_dim = ctx.gather_dim
         gather_dim = ctx.scatter_dim
+        if torch.distributed.get_rank() == 0:
+            print(f"shape before A2A: {grad_output[0].shape}")
         return_grad = _AllToAll.apply(*grad_output, process_group, scatter_dim, gather_dim)
+        if torch.distributed.get_rank() == 0:
+            print(f"shape after A2A: {return_grad.shape}")
```
Comment (ver217) [inline colossalai/shardformer/layer/_operation.py:819]: remove debug info

File: colossalai/shardformer/layer/utils.py [hunk 2]
```diff
@@ -289,3 +289,42 @@ def create_randomizer_with_offset(
...
+    Args:
+        batch (Dict[torch.Tensor]): The input tensors to split.
+        sp_group (ProcessGroup): The process group for sequence parallelism.
+
+    """
+    sp_size = dist.get_world_size(sp_group)
+    sp_rank = dist.get_rank(sp_group)
+    if sp_size > 1:
+        for key, tensor in batch.items():
+            seq_dim = 1 if key != "attention_mask" else 2
```
Comment (ver217) [inline colossalai/shardformer/layer/utils.py:310]: It seems that this function is called after embedding layer. But the seq_dim is not always 1 for hidden states. E.g. for chatglm, its seq_dim is 0.
image

Comment (Edenzzzz) [reply colossalai/shardformer/layer/utils.py]: Now provides a seq_dim argument.

File: colossalai/lazy/pretrained.py [hunk 3]
```diff
@@ -195,7 +193,6 @@ def new_from_pretrained(
                     "cache_dir": cache_dir,
                     "force_download": force_download,
                     "proxies": proxies,
-                    "resume_download": resume_download,
```
Comment (Edenzzzz) [inline colossalai/lazy/pretrained.py:198]: deprecated

**Comment by Edenzzzz** [comment]
image

File: colossalai/shardformer/layer/attn.py [hunk 4]
```diff
@@ -28,7 +37,7 @@ def invert_mask(mask: torch.Tensor) -> torch.Tensor:
     """Invert the mask tensor.
 
     Args:
-        mask (torch.Tensor): Mask tensor. Shape should be [B, 1, Sq, Skv]
+        mask (torch.Tensor): Mask tensor. Shape should be [B, 1, Sq, Sq]
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:40]: `[B, 1, Sq, Skv]` is right

File: colossalai/shardformer/layer/attn.py [hunk 5]
```diff
@@ -199,10 +227,10 @@ def attention(
...
 
         Args:
-            q (torch.Tensor): Query tensor. Shape should be [B, N, Sq, D]
-            k (torch.Tensor): Key tensor. Shape should be [B, N, Skv, D]
-            v (torch.Tensor): Value tensor. Shape should be [B, N, Skv, D]
-            attention_mask (Optional[torch.Tensor], optional): Attention mask tensor. Shape should be [B, 1, Sq, Skv]. Defaults to None.
+            q (torch.Tensor): Query tensor. Shape should be [B, nHeads, Sq, D]
+            k (torch.Tensor): Key tensor. Shape should be [B, nHeads, Skv, D]
+            v (torch.Tensor): Value tensor. Shape should be [B, nHeads, Skv, D]
+            attention_mask (Optional[torch.Tensor], optional): Attention mask tensor. Shape should be [B, 1, Sq, Sq]. Defaults to None.
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:233]: `[B, 1, Sq, Skv]` is right

File: colossalai/shardformer/layer/attn.py [hunk 6]
```diff
@@ -249,9 +277,10 @@ def attention(
                     and q_indices is not None
                     and kv_indices is not None
                 )
-        else:
-            # if attention_mask is None, attention_mask_type should be the default value
-            assert attention_mask_type == AttnMaskType.CUSTOM
+            else:
+                # if attention_mask is None, attention_mask_type should be the default value
+                assert attention_mask_type == AttnMaskType.CUSTOM
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:282]: wrong tab?

File: colossalai/shardformer/layer/attn.py (hunk 5 above)
Comment (Edenzzzz) [reply colossalai/shardformer/layer/attn.py:233]: this should be nHeads?

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -12,6 +12,7 @@ repos:
     hooks:
       - id: isort
         name: sort all imports (python)
+        args: ["--profile", "black"] # avoid comflict with black
```
[the remark this answers is not in the document]
Comment (Edenzzzz) [reply .pre-commit-config.yaml]: fixed

File: colossalai/shardformer/modeling/llama.py [hunk 8]
```diff
@@ -629,32 +662,45 @@ def forward(
...
             cache_position = torch.arange(past_seen_tokens, past_seen_tokens + seq_len, device=inputs_embeds.device)
-
         if position_ids is None:
             position_ids = cache_position.unsqueeze(0)
 
-        # in this case, attention_mask is a dict rather than a tensor
         if shard_config.enable_flash_attention:
-            mask_shape = (inputs_embeds.shape[0], 1, seq_len, past_seen_tokens + seq_len)
-            attention_mask = ColoAttention.prepare_attn_kwargs(
+            mask_shape = (batch_size, 1, past_seen_tokens + seq_len, past_seen_tokens + seq_len)
```
Comment (ver217) [inline colossalai/shardformer/modeling/llama.py:679]: `Sq` should not include `past_seen_tokens `

File: examples/language/llama/benchmark.py [hunk 9]
```diff
@@ -195,12 +207,13 @@ def empty_init():
...
             zero_stage=args.zero,
             sp_size=args.sp,
+            sequence_parallelism_mode=args.sp_mode,
             enable_sequence_parallelism=args.sp > 1,
             enable_fused_normalization=torch.cuda.is_available(),
             enable_flash_attention=args.xformers,
             microbatch_size=args.mbs,
             precision="bf16",
-            overlap_p2p=args.overlap,
+            dp_outside=False,
```
Comment (ver217) [inline examples/language/llama/benchmark.py:216]: why?

File: colossalai/shardformer/modeling/llama.py (hunk 8 above)
Comment (Edenzzzz) [reply colossalai/shardformer/modeling/llama.py]: fixed

File: examples/language/llama/benchmark.py (hunk 9 above)
Comment (Edenzzzz) [reply examples/language/llama/benchmark.py]: removed `dp_outside`

File: colossalai/shardformer/layer/attn.py [hunk 10]
```diff
@@ -144,18 +164,22 @@ def prepare_attn_kwargs(
                 attention_mask = attention_mask.tril(diagonal=0)
             attention_mask = attention_mask.expand(b, s_q, s_kv)
         else:
+            assert q_padding_mask.shape == (
+                b,
+                s_q,
+            ), f"q_padding_mask shape {q_padding_mask.shape} should be {b, s_q}."
```
[the remark this answers is not in the document]
Comment (Edenzzzz) [reply colossalai/shardformer/layer/attn.py:170]: changed to s_kv

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 11]
```diff
@@ -1108,6 +1121,8 @@ def __init__(
                 )
             else:
                 raise NotImplementedError()
+        if sequence_parallelism_mode == "ring_attn":
+            assert parallel_output, "Ring Attention doesn't support gathering output yet."
```
Comment (ver217) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:1125]: set it to false

File: colossalai/shardformer/layer/attn.py [hunk 12]
```diff
@@ -2,20 +2,30 @@
...
+from .utils import RingComm, get_half_index, split_varlen_zigzag
 
 __all__ = [
     "AttnMaskType",
     "ColoAttention",
 ]
 
+_flash_attn_forward = _flash_attn_backward = None
+_unpad_input = _pad_input = None
+logger = get_dist_logger()
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:27]: dist logger should be initialized in function

File: colossalai/shardformer/layer/attn.py [hunk 13]
```diff
@@ -144,18 +167,22 @@ def prepare_attn_kwargs(
                 attention_mask = attention_mask.tril(diagonal=0)
             attention_mask = attention_mask.expand(b, s_q, s_kv)
         else:
+            assert q_padding_mask.shape == (
+                b,
+                s_kv,
+            ), f"q_padding_mask shape {q_padding_mask.shape} should be {b, s_kv}."
```
Comment (ver217) [inline colossalai/shardformer/layer/attn.py:173]: this assertation can be removed

File: colossalai/booster/plugin/hybrid_parallel_plugin.py (hunk 11 above)
Comment (Edenzzzz) [reply colossalai/booster/plugin/hybrid_parallel_plugin.py:1125]: You mean force set to True?

File: colossalai/shardformer/layer/attn.py (hunk 13 above)
Comment (Edenzzzz) [reply colossalai/shardformer/layer/attn.py:173]: Using the existing assertion below now

File: colossalai/booster/plugin/hybrid_parallel_plugin.py (hunk 11 above)
Comment (ver217) [reply colossalai/booster/plugin/hybrid_parallel_plugin.py:1129]:
```python
if sequence_parallelism_mode == "ring_attn":
    parallel_output = False
    # raise a warning here
```

**Review by ver217 — APPROVED** [review]
